### DISCUSSION TITLE
How to approach when models output more than one tensor

### DISCUSSION THREAD

**Discussion by IZugec** [Cambridge-ICCS/FTorch#135] [Q&A] [answered]
Hello everyone,

I see that one of the open issues touches upon the topic of multiple output tensors from pytorch models https://github.com/Cambridge-ICCS/FTorch/issues/3 . I have a model which is expecting 5 tensors in the forward function and outputs 2 tensors of different shape (namely scalar, and 2D shaped tensors) Is there anything I could do at this moment to properly handle this situation? 

Thanks in advance,
Ivan

**Comment by jatkinson1000** [comment]
Thanks for asking Ivan!

This is something we have an open issue for in #3 
When we started writing this was not something libtorch could handle, so you could only return a single output tensor.
However, I think this may now be possible so I will look at implementing this.

I am currently on leave and will be back in a week, but I will also make my colleagues aware of this discussion.
Depending on your familiarity with libtorch/Fortran you would be welcome to contribute this feature yourself as this is an Open Source and Open Development project and we would be happy to guide you in this.

**Comment by IZugec** [comment]
Unfortunately I am relatively new to both Fortran and libtorch so I am not not confident I would be able to contribute meaningfully. I tried the "dumb" idea of just scripting two instances of the model, one giving scalar, the other 2D shaped tensor. The unfortunate thing is that scalar is energy and 2D array is force. One gets force as a gradient of an energy, but I ran into an another active issue https://github.com/Cambridge-ICCS/FTorch/issues/131 :(

I am not sure how complicated these problems really are. If you think they are solvable in reasonably short amount of time I would be more than happy to be guided toward general direction of what should be done, and at least try.

Thanks a lot for a quick response!
Ivan

**Reply by IZugec** [reply]
"The unfortunate thing is that scalar is energy and 2D array is force. One gets force as a gradient of an energy, but I ran into an another active issue https://github.com/Cambridge-ICCS/FTorch/issues/131"

This isn't really related to the title, apologies for that, but I would just like to expand on it. I needed gradients just in the forward pass so that I can use torch autograd to compute forces from energy. However (as I just realized), this is not a problem and can easily be achieved by defining a bool variable "requires_grad_opt" as indicated in the documentation of the torch_module_forward subroutine.

**Reply by jatkinson1000** [reply]
OK @IZugec , that's good to know :)

As for multiple outputs I have had a look and it's definitely possible, but not quite as straightforward as I had initially hoped (and my C++ isn't as good as other languages).
I have started work on this in a pull request over at #136 if you want to keen an eye on progress and will hopefully make more headway once I am back from vacation.

All best.

**Comment by jatkinson1000** [comment, marked as answer]
@IZugec We have now merged the updates to add the ability for multiple outputs.
Please check the latest version of the code, examples, and documentation to see this.

If this works for you please could you select an answer and close the discussion?
Otherwise please feel free to ask further questions.

**Comment by IZugec** [comment]
Works perfectly. Thanks a bunch for such a swift update!!
